### PR TITLE
Add a `keep_case` flag which ensures that env variables are not uppercased

### LINKED ISSUE #13713: lowercase environment variables not showing
### Checklist

- [X] I added a descriptive title
- [X] I searched open reports and couldn't find a duplicate

### What happened?

When setting environment variables in his conda environment, by using lowercase letters for the variable key, it appears conda sets the key to uppercase when it sets the variable:

(base) krishna@Krishnas-MBP ~ % conda env config vars set my_test_var="hello" -n envvartest
(base) krishna@Krishnas-MBP ~ % conda activate envvartest
(envvartest) krishna@Krishnas-MBP ~ % echo $my_test_var

(envvartest) krishna@Krishnas-MBP ~ % echo $MY_TEST_VAR
hello
(envvartest) krishna@Krishnas-MBP ~ % conda env config vars list
my_test_var = hello
Snippet from activate.py:
        # split provided environment variables into exports vs unsets
        for name, value in kwargs.items():
            if value is None:
                unset_vars.append(name.upper())
            else:
                export_vars[name.upper()] = value

### Conda Info

```shell
(base) krishna@Krishnas-MBP Desktop % conda info


     active environment : base
    active env location : /Users/krishna/anaconda3
            shell level : 1
       user config file : /Users/krishna/.condarc
 populated config files : /Users/krishna/.condarc
          conda version : 24.1.2
    conda-build version : 3.27.0
         python version : 3.11.5.final.0
                 solver : libmamba (default)
       virtual packages : __archspec=1=m1
                          __conda=24.1.2=0
                          __osx=14.2=0
                          __unix=0=0
       base environment : /Users/krishna/anaconda3  (writable)
      conda av data dir : /Users/krishna/anaconda3/etc/conda
  conda av metadata url : None
           channel URLs : https://repo.anaconda.cloud/repo/main/osx-arm64
                          https://repo.anaconda.cloud/repo/main/noarch
                          https://repo.anaconda.cloud/repo/r/osx-arm64
                          https://repo.anaconda.cloud/repo/r/noarch
                          https://repo.anaconda.cloud/repo/msys2/osx-arm64
                          https://repo.anaconda.cloud/repo/msys2/noarch
          package cache : /Users/krishna/anaconda3/pkgs
                          /Users/krishna/.conda/pkgs
       envs directories : /Users/krishna/anaconda3/envs
                          /Users/krishna/.conda/envs
               platform : osx-arm64
             user-agent : conda/24.1.2 requests/2.31.0 CPython/3.11.5 Darwin/23.2.0 OSX/14.2 solver/libmamba conda-libmamba-solver/24.1.0 libmambapy/1.5.6 aau/0.4.3 aid/0.2 c/K6Y_tgzE61NMBdNRKNUxaA s/60Vc_tQZDQxuXbiPHYHmug e/Z7_eIk9xGDoUDDdmsJvWkA
                UID:GID : 502:20
             netrc file : None
           offline mode : False


(base) krishna@Krishnas-MBP Desktop %
```

### Conda Config

_No response_

### Conda list

```shell
(base) krishna@Krishnas-MBP Desktop % conda config --show-sources

==> /Users/krishna/.condarc <==
aggressive_update_packages:
  - ca-certificates
  - certifi
  - openssl
channels:
  - defaults
default_channels:
  - https://repo.anaconda.cloud/repo/main
  - https://repo.anaconda.cloud/repo/r
  - https://repo.anaconda.cloud/repo/msys2
restore_free_channel: False
```

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by ForgottenProgramme**
### Description
resolves #13713 

At present `conda` converts all environment variables into uppercase. 
This PR adds a `keep_case` context variable, which allows users to store their environment variables as is.

When the `--keep-case` flag is used, environment variables are saved in "preserved case". When the flag is not used variables are saved in upper case (the original conda behavior). 

**Tasks**

- [x] Add a new `keep_case` variable to context
- [x] store environment variables in preserved case when `keep_case` is `True`
- [x] write tests
- [x] write more tests
- [x] Parametrize the tests!

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [ ] Add / update outdated documentation?

### CODE REVIEW

**Comment by ForgottenProgramme** [comment]
pre-commit.ci autofix

File: tests/test_activate.py [hunk 1]
```diff
@@ -3355,6 +3357,126 @@ def test_activate_and_deactivate_for_uninitialized_env(conda_cli):
...
+    assert "one" not in export_vars
+    assert "three" not in unset_vars
+
+    # to uppercase
+    assert "ONE" in export_vars
+    assert "THREE" in unset_vars
+
+    # original uppercase
+    assert "TWO" in export_vars
+    assert "FOUR" in unset_vars
```
Comment (kenodegard) [inline tests/test_activate.py:3409]: Combine both of these with parameterization:

```suggestion
@pytest.mark.parametrize("keep_case", [True, False])
def test_keep_case(monkeypatch: MonkeyPatch, keep_case: bool):
    monkeypatch.setenv("CONDA_KEEP_CASE", keep_case)
    reset_context()
    assert context.keep_case is keep_case

    activator = PosixActivator()
    export_vars, unset_vars = activator.get_export_unset_vars(
        one=1,
        TWO=2,
        three=None,
        FOUR=None,
    )

    # original lowercase (conditionally true per keep_case)
    assert ("one" in export_vars) is keep_case
    assert ("three" in unset_vars) is keep_case

    # to uppercase (always true)
    assert "ONE" in export_vars
    assert "THREE" in unset_vars

    # original uppercase (always true)
    assert "TWO" in export_vars
    assert "FOUR" in unset_vars
```

File: tests/test_activate.py [hunk 2]
```diff
@@ -3355,6 +3357,126 @@ def test_activate_and_deactivate_for_uninitialized_env(conda_cli):
...
+    assert "ONE" in export_vars
+    assert "FOUR" in unset_vars
+
+    # original lowercase
+    assert "two" not in export_vars
+    assert "three" not in unset_vars
+
+    # to uppercase
+    assert "TWO" in export_vars
+    assert "THREE" in unset_vars
```
Comment (kenodegard) [inline tests/test_activate.py:3477]: Do the same here

File: conda/activate.py [hunk 3]
```diff
@@ -115,18 +115,30 @@ def get_export_unset_vars(self, export_metavars=True, **kwargs):
...
             for name, value in context.conda_exe_vars_dict.items():
                 if value is None:
                     unset_vars.append(name.upper())
+                    if context.keep_case:
+                        unset_vars.append(name)
+
                 elif "/" in value or "\\" in value:
                     export_vars[name.upper()] = self.path_conversion(value)
+                    if context.keep_case:
+                        export_vars[name] = self.path_conversion(value)
```
Comment (kenodegard) [inline conda/activate.py:137]: This clause is untested in `test_metavars_keep_case`. Let's adjust the mocked value for `context.conda_exe_vars_dict` to include a path-like value.

**Review by kenodegard — CHANGES_REQUESTED** [review]

**Review by kenodegard — CHANGES_REQUESTED** [review]

File: tests/test_activate.py [hunk 4]
```diff
@@ -3355,6 +3357,90 @@ def test_activate_and_deactivate_for_uninitialized_env(conda_cli):
...
+    if not keep_case_boolean:
+        assert not context.keep_case
+
+    returned_dict = {
+        "ONE": "1",
+        "two": "2",
+        "three": None,
+        "FOUR": None,
+        "five": "a/path/to/something",
+        "SIX": "a\\path",
```
Comment (kenodegard) [inline tests/test_activate.py:3413]: `SIX` is not included in the assertions below

**Comment by jezdez** [comment]
Close in favor of #13943.

**Pull request closed without merging**